allennlp/modules/transformer/bert_biattention.py

```diff
@@ -0,0 +1,122 @@
+import math
+
+import torch
+
+from allennlp.common import FromParams
+
+
+class BertBiAttention(torch.nn.Module, FromParams):
+    def __init__(
+        self,
+        hidden_size1: int,
+        hidden_size2: int,
+        combined_hidden_size: int,
+        num_attention_heads: int,
+        dropout1: float,
+        dropout2: float,
+    ):
+        super().__init__()
+        if combined_hidden_size % num_attention_heads != 0:
+            raise ValueError(
+                "The hidden size (%d) is not a multiple of the number of attention "
+                "heads (%d)" % (combined_hidden_size, num_attention_heads)
+            )
+
+        self.num_attention_heads = num_attention_heads
+        self.attention_head_size = int(combined_hidden_size / num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        self.query1 = torch.nn.Linear(hidden_size1, self.all_head_size)
+        self.key1 = torch.nn.Linear(hidden_size1, self.all_head_size)
+        self.value1 = torch.nn.Linear(hidden_size1, self.all_head_size)
+
+        self.dropout1 = torch.nn.Dropout(dropout1)
+
+        self.query2 = torch.nn.Linear(hidden_size2, self.all_head_size)
+        self.key2 = torch.nn.Linear(hidden_size2, self.all_head_size)
+        self.value2 = torch.nn.Linear(hidden_size2, self.all_head_size)
+
+        self.dropout2 = torch.nn.Dropout(dropout2)
+
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size,)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
```

Comment:
This should either be documented, or (more likely) prefixed with a `_`. I'm assuming this is an internal function, right?